### PR TITLE
Migrate CLI to Click, add Version Upgrade Logic

### LINKED ISSUE #5675: Unable to start kolibri after upgrading to 0.12.4
### Observed behavior

Reported by a user:
cannot start kolibri after upgrading it from 0.12.3 to 0.12.4. The file `.data_version` in the kolibri home directory does not contain the version 0.12.3 but has some empty spaces.
backup_path here was something like `db-v       _<dtm>.dump`
Screen Shot 2019-06-18 at 11 39 35 AM

### Expected behavior

kolibri should start successfully 

### User-facing consequences

users are not able to use kolibri

### Errors and logs

…

### Steps to reproduce

kolibri.txt:
```
INFO 2019-05-13 19:37:32,433 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 19:37:33,470 cli Kolibri running for the first time.
INFO 2019-05-13 19:37:33,470 cli We don't yet use pre-migrated database seeds, so you're going to have to wait a bit while we create a blank database...
INFO 2019-05-13 19:37:57,103 cli Automatically enabling applications.
INFO 2019-05-13 19:37:57,104 cli Running update routines for new version...
INFO 2019-05-13 19:38:05,354 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 19:38:05,376 cli Running 'kolibri start' as daemon (system service)
ERROR 2019-05-13 19:38:05,608 cli Could not detect an IP address that Kolibri binds to, but try opening up the following addresses:
INFO 2019-05-13 19:38:05,694 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 19:38:05,731 ping Attempting a ping.
INFO 2019-05-13 19:38:06,223 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 19:38:06,223 vacuumsqlite Next Vacuum at 3AM local server time (in 441.8833333333333 minutes).
WARNING 2019-05-13 19:38:06,445 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 19:38:06,923 utils Started new MESSAGEPROCESSOR thread ID#sad-violet-washington-aspen
INFO 2019-05-13 19:38:06,923 utils Started new SCHEDULER thread ID#sierra-tennessee-sweet-december
INFO 2019-05-13 19:38:06,924 utils Started new WORKERMESSAGEHANDLER thread ID#white-fillet-mountain-xray
INFO 2019-05-13 19:38:07,057 wsgi Starting Kolibri
INFO 2019-05-13 14:52:38,061 ping Attempting a ping.
INFO 2019-05-13 14:52:40,799 ping Ping succeeded! (response: {'messages': [], 'id': 231251})
INFO 2019-05-13 14:52:43,500 ping Sleeping for 1440.0 minutes.
INFO 2019-05-13 16:56:03,431 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 16:56:05,197 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 16:56:05,298 cli Running update routines for new version...
INFO 2019-05-13 16:56:08,574 cli Running 'kolibri start' as daemon (system service)
INFO 2019-05-13 16:56:08,744 cli Kolibri running on:
INFO 2019-05-13 16:56:08,744 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 16:56:08,790 ping Attempting a ping.
INFO 2019-05-13 16:56:09,207 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 16:56:09,207 vacuumsqlite Next Vacuum at 3AM local server time (in 603.8333333333334 minutes).
WARNING 2019-05-13 16:56:09,477 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 16:56:09,523 utils Started new MESSAGEPROCESSOR thread ID#mars-asparagus-south-mango
INFO 2019-05-13 16:56:09,523 utils Started new SCHEDULER thread ID#kentucky-idaho-glucose-burger
INFO 2019-05-13 16:56:09,539 utils Started new WORKERMESSAGEHANDLER thread ID#west-december-pizza-high
INFO 2019-05-13 16:56:09,646 wsgi Starting Kolibri
INFO 2019-05-13 17:13:26,919 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 17:13:28,403 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:13:28,466 cli Running update routines for new version...
INFO 2019-05-13 17:13:31,590 cli Running 'kolibri start' as daemon (system service)
ERROR 2019-05-13 17:13:31,793 cli Could not detect an IP address that Kolibri binds to, but try opening up the following addresses:
INFO 2019-05-13 17:13:32,434 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 17:13:32,684 ping Attempting a ping.
INFO 2019-05-13 17:13:33,309 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:13:33,309 vacuumsqlite Next Vacuum at 3AM local server time (in 586.4333333333333 minutes).
WARNING 2019-05-13 17:13:33,465 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 17:13:33,871 utils Started new MESSAGEPROCESSOR thread ID#fruit-triple-nevada-avocado
INFO 2019-05-13 17:13:33,871 utils Started new SCHEDULER thread ID#georgia-neptune-avocado-sink
INFO 2019-05-13 17:13:33,871 utils Started new WORKERMESSAGEHANDLER thread ID#spring-berlin-violet-south
INFO 2019-05-13 17:13:34,043 wsgi Starting Kolibri
INFO 2019-05-13 17:28:33,484 ping Attempting a ping.
WARNING 2019-05-13 17:28:33,584 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 17:38:36,138 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 17:38:37,450 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:38:37,512 cli Running update routines for new version...
INFO 2019-05-13 17:38:40,074 cli Running 'kolibri start' as daemon (system service)
ERROR 2019-05-13 17:38:40,168 cli Could not detect an IP address that Kolibri binds to, but try opening up the following addresses:
INFO 2019-05-13 17:38:40,246 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 17:38:40,277 ping Attempting a ping.
INFO 2019-05-13 17:38:40,793 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:38:40,793 vacuumsqlite Next Vacuum at 3AM local server time (in 561.3166666666667 minutes).
WARNING 2019-05-13 17:38:40,902 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 17:38:40,996 utils Started new MESSAGEPROCESSOR thread ID#cold-triple-beer-double
INFO 2019-05-13 17:38:40,996 utils Started new SCHEDULER thread ID#salami-moon-twenty-uniform
INFO 2019-05-13 17:38:40,996 utils Started new WORKERMESSAGEHANDLER thread ID#carpet-one-carpet-ohio
INFO 2019-05-13 17:38:41,105 wsgi Starting Kolibri
INFO 2019-05-13 17:42:32,446 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 17:42:33,946 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:42:34,039 cli Running update routines for new version...
INFO 2019-05-13 17:42:37,570 cli Running 'kolibri start' as daemon (system service)
ERROR 2019-05-13 17:42:37,663 cli Could not detect an IP address that Kolibri binds to, but try opening up the following addresses:
INFO 2019-05-13 17:42:37,741 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 17:42:37,757 ping Attempting a ping.
INFO 2019-05-13 17:42:38,101 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:42:38,101 vacuumsqlite Next Vacuum at 3AM local server time (in 557.35 minutes).
WARNING 2019-05-13 17:42:38,382 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 17:42:38,507 utils Started new MESSAGEPROCESSOR thread ID#double-pasta-sixteen-jersey
INFO 2019-05-13 17:42:38,507 utils Started new SCHEDULER thread ID#helium-delta-oscar-bakerloo
INFO 2019-05-13 17:42:38,507 utils Started new WORKERMESSAGEHANDLER thread ID#georgia-apart-lemon-music
INFO 2019-05-13 17:42:38,601 wsgi Starting Kolibri
INFO 2019-05-13 17:45:17,563 apps Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO 2019-05-13 17:45:19,375 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:45:19,519 cli Running update routines for new version...
INFO 2019-05-13 17:45:22,807 cli Running 'kolibri start' as daemon (system service)
ERROR 2019-05-13 17:45:22,885 cli Could not detect an IP address that Kolibri binds to, but try opening up the following addresses:
INFO 2019-05-13 17:45:22,979 cli Going to daemon mode, logging to C:\Users\Specktron\.kolibri\server.log
INFO 2019-05-13 17:45:23,010 ping Attempting a ping.
INFO 2019-05-13 17:45:23,370 vacuumsqlite Sqlite database Vacuum finished.
INFO 2019-05-13 17:45:23,370 vacuumsqlite Next Vacuum at 3AM local server time (in 554.6 minutes).
WARNING 2019-05-13 17:45:23,494 ping Ping failed (could not connect). Trying again in 15.0 minutes.
INFO 2019-05-13 17:45:23,635 utils Started new MESSAGEPROCESSOR thread ID#indigo-carbon-minnesota-paris
INFO 2019-05-13 17:45:23,635 utils Started new SCHEDULER thread ID#potato-uncle-london-double
INFO 2019-05-13 17:45:23,635 utils Started new WORKERMESSAGEHANDLER thread ID#arkansas-orange-potato-cat
INFO 2019-05-13 17:45:23,744 wsgi Starting Kolibri
WARNING 2019-06-17 07:44:05,567 conf Default plugin kolibri.plugins.default_theme not found in configuration. To re-disable it, run:
   $ kolibri plugin kolibri.plugins.default_theme disable
```

notice that the log message `INFO 2019-05-13 17:45:23,744 wsgi Starting Kolibri` does not include the kolibri version.

### Context

kolibri version: 0.12.3 --> 0.12.4
operating system: Windows

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
* Migrates the Kolibri CLI to click
* Adds the ability to do version constrained upgrade logic
* Cleans up accumulated CLI cruft to restrict startup tasks.

### Reviewer guidance
Does the CLI work as expected?
Does the version upgrade logic make sense, is it sufficiently well documented?

### References
Fixes #5675

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/utils/version.py [hunk 1]
```diff
@@ -103,6 +103,8 @@
 import re
 import subprocess
 
+import semver
```
Comment (rtibbles) [inline kolibri/utils/version.py]: This was an error. I should either import this inside the functions that use it, or define them external to this file.

File: kolibri/core/content/upgrade.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            if not ChannelMetadata.objects.filter(id=channel_id).exists():
+                try:
+                    import_channel_from_local_db(channel_id)
+                    annotate_content(channel_id)
+                except (InvalidSchemaVersionError, FutureSchemaError):
+                    logger.warning(
+                        "Tried to import channel {channel_id}, but database file was incompatible".format(
+                            channel_id=channel_id
+                        )
+                    )
```
Comment (indirectlylit) [inline kolibri/core/content/upgrade.py:55]: This is not a blocking comment, but I'll add my usual inheritance-averse reaction:

I think this logic would be better represented as instances of a `VersionUpgrade` class rather than subclasses of a `VersionUpgradeBase` class.

The main reason is that with subclasses, there's an implicit assumption that subclass can "extend" the API or change behaviors as desired. Having class instances instead makes the API and contract fully defined and immutable, which as far as I can tell would better serve the purpose of the `run_upgrades` function.

For example:

```python
from kolibri.core.upgrade import VersionUpgrade

def importExternalContentDBs(self):
    """
    If we are potentially moving from a version of Kolibri that did not import its content data,
    scan through the content database folder for all channel content databases,
    and pull the data from each database if we have not already imported it.
    """
    channel_ids = get_channel_ids_for_content_database_dir(
        get_content_database_dir_path()
    )
    for channel_id in channel_ids:
        if not ChannelMetadata.objects.filter(id=channel_id).exists():
            try:
                import_channel_from_local_db(channel_id)
                annotate_content(channel_id)
            except (InvalidSchemaVersionError, FutureSchemaError):
                logger.warning(
                    "Tried to import channel {channel_id}, but database file was incompatible".format(
                        channel_id=channel_id
                    )
                )

def getUpgrades():
    return [
        VersionUpgrade(func=importExternalContentDBs, old_version="<0.6.0"),
    ]
```

Comment (rtibbles) [reply kolibri/core/content/upgrade.py:55]: Fair comment - I just thought I'd get away with it because I restrained myself from using metaclasses :)

Comment (rtibbles) [reply kolibri/core/content/upgrade.py:55]: Have now refactored to rely on instantiation, with a convenience decorator to just simply apply it to a function to make it into an upgrade operation.

**Comment by rtibbles** [comment]
Getting the `--` methodology to work after this refactor will be very challenging, but given that we removed any documentation of this before the release of 0.5.0, I am not sure we have any reason to continue to honour this.

**Comment by indirectlylit** [comment]
> Getting the -- methodology to work 

What are the main use cases when we needed that?

**Comment by rtibbles** [comment]
In the earliest versions of Kolibri it was required to allow us to distinguish kolibri args from django management command args, but then it was dropped and completely deprecate in 0.5.0 - but we maintained backwards compatibility, but not sure who needed it.

**Comment by indirectlylit** [comment]
it looks like we currently have that syntax [documented](https://kolibri.readthedocs.io/en/latest/manage.html#import-content-channels-from-internet), e.g.:

```
kolibri manage importchannel -- network <Channel ID>
kolibri manage importcontent -- network <Channel ID>
```

Perhaps this is a good time to reconsider the proposal I made [here](https://github.com/learningequality/kolibri/issues/3727#issuecomment-394761560) that our Django management commands should be considered an internal implementation detail, and the user-facing UI could be simplified.

Some examples:

| Current Django command | Possible alternative Kolibri command |
|--|--|
| `kolibri manage listchannels` | `kolibri channels` |
| `kolibri manage importchannel -- network <channel>`<br>`kolibri manage importcontent -- network <channel>` | `kolibri channels <channel> --import` |
| `kolibri manage setchannelposition <channel> <pos>` | `kolibri channels <channel> --set-pos <pos>` |
| `kolibri manage createsuperuser` | `kolibri users --create-super-admin` |
| `kolibri manage importusers your-csv-file.csv` |  `kolibri users --import --csv your-csv-file.csv` |
| `kolibri manage deleteuser <username>` | `kolibri users <username> --delete` |
| `kolibri manage content movedirectory <destination>` | `kolibri config --content-directory <destination>` |
| ... | ... |

**Comment by rtibbles** [comment]
OK, this is rebased, and I think should be ready for final review.

Following comments from @indirectlylit above, and from our direct conversations, I have removed the attempts to introspect Django management commands with Click, and instead am now just doing a simple pass through, follow up work can implement some of the pieces that @indirectlylit suggests above. My vote would be to move most of our logic out of the management commands themselves, so that it can be invoked by one of three things:
1) A Django management command (for full backwards compatibility)
2) A Click command (as suggested above)
3) An iceqube or rq task wrapper
At the moment we are conflating 1) and 3), which would be good to unpick for better job parallelization.

**Comment by rtibbles** [comment]
Oh, and the current implementation simply ignores `--`. This is not compatible with separately passing the arguments to kolibri ahead of `--`, but at least provides compatibility with previously documented examples:
```
kolibri manage importchannel -- network <Channel ID>
kolibri manage importcontent -- network <Channel ID>
```

**Comment by rtibbles** [comment]
Oh - this isn't quite ready for merge yet, because all the global flags are currently only exposed on the main kolibri command, resolving that now.

**Comment by rtibbles** [comment]
This should now be resolved.

File: docs/backend_architecture/upgrade.rst [hunk 3]
```diff
@@ -18,7 +18,31 @@ and create a backup before doing so.
...
+``version_upgrade``. An toy example is shown below.
+
+.. code-block:: python
+
+  import logging
+  from kolibri.core.upgrade import version_upgrade
+
+  logger = logging.getLogger(__name__)
+
+  @version_upgrade(old_version="<0.6.4", new_version=">=1.0.0")
```
Comment (indirectlylit) [inline docs/backend_architecture/upgrade.rst:46]: To clarify, are these "database versions" or Kolibri versions?

I think the rest of the content on this page is framed as user-facing instructions talking about Kolibri versions rather than developer docs talking about the `data_version`.

Comment (rtibbles) [reply docs/backend_architecture/upgrade.rst:46]: These are Kolibri versions.

I found the inclusion of these instructions in our dev docs confusing, actually, because they did seem to be user facing, rather than dev facing.

Comment (indirectlylit) [reply docs/backend_architecture/upgrade.rst:46]: yes. I think they could use a re-work, because I'm not sure they're actually accurate.

Would you mind giving the the other stuff in there a quick look and remove if appropriate?

Comment (rtibbles) [reply docs/backend_architecture/upgrade.rst:46]: I removed everything, and then added some bonus material about django schema migrations.

**Review by indirectlylit — APPROVED** [review]
Made some tweaks:

* tightened up user-facing strings
* switched to using `help=` rather than comment parsing
* switched `--daemon` to `--background`, while keeping support for the former but avoiding having it displayed in the help
* switched `--skipupdates` to `--skip-updates` even though "ski pup dates" sound awesome. Did not include backward compatibility because I couldn't find this option documented anywhere.

Assuming these look good to you and things pass, feel free to merge.

A few follow-up observations:

* The `kolibri plugin enable`/`disable` command and sub-commands have poor help text. Might be cleaner to split it into two top-level commands (`kolibri enable-plugin`/`disable-plugin`) or do it as options (`kolibri plugin --enable`/`--disable`)
* In many cases (like that one) I anticipate we'll want a robust, maintainable way of supporting old command forms without exposing them in the main help text
* Eventually, it might be nice to hide or separate the "advanced" commands like `manage` and `shell` from the more common commands like `start` and `stop`

**Comment by rtibbles** [comment]
I think the plugin command needs to be improved over all, and we have a few use cases where we could look at how to enhance this API.

File: kolibri/utils/cli.py [hunk 4]
```diff
@@ -333,11 +333,18 @@ def create_startup_lock(port):
         logger.warn(u"Impossible to create file lock to communicate starting process")
 
 
-@main.command(
-    cls=KolibriDjangoCommand,
-    help=u"Start the Kolibri process",
-    context_settings=dict(ignore_unknown_options=True, allow_extra_args=True),
-)
+def get_deprecated_daemon_arg(background):
```
Comment (indirectlylit) [inline kolibri/utils/cli.py]: looks like you forgot to remove this function

Comment (rtibbles) [reply kolibri/utils/cli.py]: D'oh!

File: kolibri/utils/cli.py [hunk 5]
```diff
@@ -1,17 +1,25 @@
+"""
+We do not import unicode_literals here as it causes Click to
+give some very aggressive warning errors about subtle bugs.
+So, to avoid these errors and these subtle bugs, we use
+explicit unicode literals in the following.
+"""
 from __future__ import absolute_import
 from __future__ import print_function
-from __future__ import unicode_literals
```
Comment (benjaoming) [inline kolibri/utils/cli.py]: Wondering... Why was this removed and `u"string"` brought back? CC: @rtibbles

Comment (benjaoming) [inline kolibri/utils/cli.py:3]: Sorry, first I commented on this without reading the comment above :)

I've now seen these two sources:

1. https://github.com/PythonCharmers/python-future/issues/22
1. https://click.palletsprojects.com/en/7.x/python3/#unicode-literals

I understand that the motivation from discouraging this in `python-future` and `Click` doesn't have anything to do with Click specifically.

Since we have this import throughout the project, IMO it's fine. Most of the historic issues of why Python 2.7 had unicode issues have been fixed, and Python 2.7 support in itself isn't really encouraged.

I think it's okay to re-introduce this, considering that everything else in the project wants unicode for Python 2.7.

If you agree, I can open a new issue. CC: @rtibbles

Comment (rtibbles) [reply kolibri/utils/cli.py:3]: The main issue is that Click spews a huge number of warnings for anything we do through the CLI if we don't use unicode literals - or are you suggesting to just reintroduce the import block and maintain the `u` literals too? (I actually can't remember which one triggers the warnings from `Click`.

Comment (benjaoming) [reply kolibri/utils/cli.py:3]: They advice not to do this:

```
import click
click.disable_unicode_literals_warning = True
```

...but that could be done if we feel comfortable about it. I don't quite understand if there are issues pertaining unicode and Python 2.7 for Click, or if they are just doing this as some general discouragement. From their own docs and the referred issue, I don't see any reasoning that isn't just general advice, and I think the consequence of agreeing to this, would be to refactor our whole codebase to be explicitly unicode `u"string"`.

Using unicode literals in Python 2.7+ has always been a way to get Py2+3 support in the same codebase. Relying on explicit unicode probably introduces far more issues (because humans forget) than the opposite.

Comment (indirectlylit) [reply kolibri/utils/cli.py:3]: that reasoning makes sense to me - I'd personally be OK with setting `click.disable_unicode_literals_warning = True` for the sake of string consistency.

@benjaoming comments on PRs that are already merged are liable to get lost. If you feel that this something that should be addressed, please open a follow-up PR or issue

Comment (benjaoming) [reply kolibri/utils/cli.py:3]: @indirectlylit choosing the code comment because I needed clarity before opening an issue/PR

> If you agree, I can open a new issue. CC: @rtibbles

Will do that!

Thanks for weighing in both! :heart:

**Comment by radinamatic** [comment]
@indirectlylit Was the label for `user docs` or the `dev docs`...? 🤔

**Comment by indirectlylit** [comment]
User docs. There might (?) be some subtle differences with some of the command-line behaviors. @rtibbles will be able to advise once he's feeling better

**Comment by rtibbles** [comment]
I think the main thing is that the `--` in the middle of django management commands is now completely removed and no longer supported, so we should update any docs that include that.
e.g.: `kolibri manage importchannel -- network <channel_id>`
should be changed to `kolibri manage importchannel network <channel_id>`
